### PR TITLE
101 be notification endpoint

### LINKED ISSUE #101: [BE] notification endpoint
# **Retrieve All User's Notifications Endpoint**

---

## **Acceptance Criteria**
GET /users/notifications: Get notifications that user receives with an additional field url that is relevant to notification (as in Instagram) to display a photo associated with notification on frontend. 

To do this, the notification service will compose Media Service (which composes S3 Service) to retrieve presigned url using objectKey associated with notification. For example:
- If a notification is a someone liked my post, someone commented on a post or someone liked my comments, we query database for user's profilePhoto (which is the object key) and use it to get pre-signed url of profilePhoto
- If a notification is a post, we query database to retrieve the objectKey of the first photo of the post in media table and use that objectKey to get pre-signed url of that photo

> [!NOTE]
> - The endpoint should be paginated with limit and page (a.k.a. offset)
> - Profile photo in usersTable is objectKey, and objectKey is used to get pre-signed url using Media Service and S3 client 
> - Can modify MediaService interface and method as needed
> - Will need to seed like, comment, post, and notification for testing
<br> 

- [x] Define a schema and endpoints in the openapi.yaml schema.
- [x] Define the route for notifications and create the proper type aliases from the schema generation.
- [x] Define notification transaction interface and its implementation to work with the database.
- [ ] Define notification service interface to properly handle core business logic together as well as delegate to transaction layer as well as taking in MediaService and use it when necessary.
- [ ] Define notification controller interface, properly type checking with the previously defined type aliases from schema. <br>

[Backend Guidelines](https://github.com/GenerateNU/dearly/blob/main/docs/BACKEND.md)
[Database Drizzle Schema](https://github.com/GenerateNU/dearly/blob/main/backend/src/entities/schema.ts)
[Database Supabase Visualization](https://supabase.com/dashboard/project/wtmsxzaucgdwttgwpmjy/database/schemas)

---

## **Testing Plan**

- Brainstorm different scenarios and edge cases and what kind of HTTP status code would be returned in each one.
- Ensure that the controller returns the proper response codes depending on proper requests.
- Ensure that the API returns the exact schema as defined in the OpenAPI.yaml file, this includes error responses as well.
- Testing a full cycle would also be great as well.

---

## **Definition of Done**

1. All acceptance criteria are met.  
2. All tests pass locally and cover all edge cases.

### PR DESCRIPTION
**Pull request by aahiltn**
# Description

Added notification endpoints, including getting the signed url from S3. For sake of simplicity, the logic to get the signedURL is done in the transaction, with the objectKey changing depending on the type of notification.

LikeComment is also removed from the yaml for a notification, and the relation is removed between likeCommentTable and notificationsTable.

## Checklist:

- [ ] I have performed a self-review of my code.
- [ ] I have added necessary tests (unit, integration, etc.) to cover the new functionality or fix.
- [ ] All new and existing tests pass locally.
- [ ] I have updated the documentation to reflect changes (e.g., README, code comments).
- [ ] I have removed any commented-out code, debug statements, or unnecessary console logs.

closes #101

### CODE REVIEW

File: backend/src/tests/helpers/test-constants.ts [hunk 1]
```diff
@@ -168,3 +172,48 @@ export const COMMENTS: (typeof commentsTable.$inferInsert)[] = [
...
+    title: "New Comment",
+    description: "Alice commented on your post",
+    createdAt: new Date(),
+    groupId: DEARLY_GROUP_ID,
+    commentId: DEARLY_COMMENT_ID,
+    likeId: null,
+    mediaURL: MOCK_SIGNED_URL,
+  },
+  {
+    id: MOCK_LIKE_NOTIF_UUID,
```
Comment (in-mai-space) [inline backend/src/tests/helpers/test-constants.ts:192]: for the like and post notification, we will have to seed post (there are probably post already so we can use UUID of those posts maybe) and like (will need to seed these since notification will try to identify foreign key to like table but no like exists). i think it should stop throwing foreign key constraints error when you do this!

File: backend/src/tests/users/notifications.test.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        },
+        autoAuthorized: false,
+        headers: {
+          Authorization: "Bearer invalid.token",
+        },
+      })
+    ).assertStatusCode(Status.Unauthorized);
+  });
+
+  it("should return 400 when query parameters are invalid", async () => {
```
Comment (in-mai-space) [inline backend/src/tests/users/notifications.test.ts:88]: can we also test for negative value for limit and page, and also check if pagination works with different page and limit?

**Review by in-mai-space — APPROVED** [review]
LGTM 💥🚀🐐
